File: algorithms_dataStructures/quick-sort.py
def quicksort(A, p, r):
    if p < r:
        q = partition(A, p, r)
        quicksort(A, p, q - 1)
        quicksort(A, q + 1, r)


def partition(A, p, r):
    x = A[r]
    i = p - 1
    for j in range(p, r, 1):
        if A[j] <= x:
            i += 1
            temp = A[i]
            A[i] = A[j]
            A[j] = temp
    temp = A[i + 1]
    A[i + 1] = A[r]
    A[r] = temp
    return i + 1


if __name__ == '__main__':
    a = [88, 31, 52, 25, 98, 14, 30, 62, 23, 79]
    quicksort(a, 0, len(a) - 1)
    print(a)

File: Golang/go-6/27-data-structure-and-algorithms/quick-sort.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func main() {

	slice := generateSlice(20)
	fmt.Println("\n--- Unsorted --- \n\n", slice)
	quickSort(slice)
	fmt.Println("\n--- Sorted ---\n\n", slice, "\n")
}

// Generates a slice of size, size filled with random numbers
func generateSlice(size int) []int {

	slice := make([]int, size, size)
	rand.Seed(time.Now().UnixNano())
	for i := 0; i < size; i++ {
		slice[i] = rand.Intn(999) - rand.Intn(999)
	}
	return slice
}

func quickSort(a []int) []int {
	if len(a) < 2 {
		return a
	}
	left, rigth := 0, len(a)-1
	pivot := rand.Int() % len(a)

	a[pivot], a[rigth] = a[rigth], a[pivot]
	for i, _ := range a {
		if a[i] < a[rigth] {
			a[left], a[i] = a[i], a[left]
			left++
		}
	}
	a[left], a[rigth] = a[rigth], a[left]

	quickSort(a[:left])
	quickSort(a[left+1:])

	return a
}
